File: backend/app/api/v2/endpoints/simulations.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload
from uuid import UUID

from backend.app.api.deps import get_db
from backend.app.api.schemas.simulation import (
    SimulationRunRequest,
    SimulationRunResponse,
    SimulationJobStatusResponse,
    SimulationResultData,
)
from backend.app.domains.simulation.job_repository import SimulationJobRepository
from backend.app.domains.simulation.tasks import run_monte_carlo_simulation
from backend.app.models.strategy import SimulationJob

router = APIRouter()


@router.post("/run", response_model=SimulationRunResponse)
async def run_simulation(
    body: SimulationRunRequest,
    db: AsyncSession = Depends(get_db),
):
    job_repo = SimulationJobRepository(db)
    job = SimulationJob(params=body.model_dump())
    job = await job_repo.create(job)
    await db.commit()

    run_monte_carlo_simulation.delay(str(job.id), body.model_dump())

    return SimulationRunResponse(job_id=job.id, status=job.status.value)


@router.get("/{job_id}", response_model=SimulationJobStatusResponse)
async def get_simulation_job(
    job_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(SimulationJob)
        .where(SimulationJob.id == job_id)
        .options(selectinload(SimulationJob.result))
    )
    job = result.scalar_one_or_none()

    if not job:
        raise HTTPException(status_code=404, detail="Job not found")

    result_data = None
    if job.result_id and job.result:
        r = job.result
        result_data = SimulationResultData(
            win_prob=r.win_prob,
            podium_prob=r.podium_prob,
            avg_finish=r.avg_finish,
            best_case=r.best_case,
            worst_case=r.worst_case,
            iteration_count=r.iteration_count,
            execution_time_ms=r.execution_time_ms,
        )

    return SimulationJobStatusResponse(
        job_id=job.id,
        status=job.status.value,
        created_at=job.created_at,
        completed_at=job.completed_at,
        result=result_data,
        error_message=job.error_message,
    )

File: frontend/src/api/simulations.ts
const API_BASE = import.meta.env.VITE_API_URL || "http://127.0.0.1:8000";

export interface SimulationRunRequest {
  circuit_id: string;
  driver_id: string;
  compound: string;
  tyre_age: number;
  laps_remaining: number;
  gap_ahead?: number;
  gap_behind?: number;
  base_position?: number;
  iterations?: number;
}

export interface SimulationRunResponse {
  job_id: string;
  status: string;
}

export interface SimulationResultData {
  win_prob: number;
  podium_prob: number;
  avg_finish: number;
  best_case: number;
  worst_case: number;
  iteration_count: number;
  execution_time_ms: number;
}

export interface SimulationJobStatusResponse {
  job_id: string;
  status: string;
  created_at: string;
  completed_at: string | null;
  result: SimulationResultData | null;
  error_message: string | null;
}

export async function runSimulation(
  params: SimulationRunRequest,
): Promise<SimulationRunResponse> {
  const res = await fetch(`${API_BASE}/api/v2/simulations/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!res.ok) throw new Error(`Simulation run error [${res.status}]`);
  return res.json();
}

export async function getSimulationJob(
  jobId: string,
): Promise<SimulationJobStatusResponse> {
  const res = await fetch(`${API_BASE}/api/v2/simulations/${jobId}`);
  if (!res.ok) throw new Error(`Simulation job status error [${res.status}]`);
  return res.json();
}
